calico_containers/calicoctl.py

```diff
@@ -664,67 +668,82 @@ def save_diags():
-echo "Using temp dir: $diags_dir"
-pushd $diags_dir >/dev/null
+    temp_dir = tempfile.mkdtemp()
+    temp_diags_dir = os.path.join(temp_dir, 'diagnostics')
+    os.mkdir(temp_diags_dir)
 
-echo DATE=$date > date
-echo $system > hostname
+    print("Using temp dir: %s" % temp_dir)
 
-echo "Dumping netstat output"
-netstat -an > $diags_dir/netstat
+    with open(os.path.join(temp_diags_dir, 'date'), 'w') as f:
+        f.write("DATE=%s" % datetime.strftime(datetime.today(),"%Y-%m-%d_%H-%M-%S"))
 
-echo "Dumping routes"
-for cmd in "route -n" "ip route" "ip -6 route"
-do
-  echo $cmd >> $ROUTE_FILE
-  $cmd >> $ROUTE_FILE
-  echo >> $ROUTE_FILE
-done
-netstat -an > netstat
+    with open(os.path.join(temp_diags_dir, 'hostname'), 'w') as f:
+        f.write("%s" % socket.gethostname())
 
-echo "Dumping iptables"
-iptables-save > $IPTABLES_PREFIX
-ipset list > ipset
+    print("Dumping netstat output")
+    with open(os.path.join(temp_diags_dir, 'netstat'), 'w') as f:
+        subprocess.Popen(["netstat", "-an"], stdin=subprocess.PIPE, stdout=f)
 
-echo "Copying Calico logs"
-cp -a $CALICO_DIR .
+    print("Dumping routes")
+    with open(os.path.join(temp_diags_dir, 'route'), 'w') as f:
+        # Could do this in a for loop...
+        f.write("route -n")
+        subprocess.Popen(["route", "-n"], stdin=subprocess.PIPE, stdout=f)
+        f.write('\n')
 
-echo "Dumping datastore"
-curl -s -L http://127.0.0.1:4001/v2/keys/calico?recursive=true -o etcd_calico
+        f.write("ip route")
+        subprocess.Popen(["ip", "route"], stdin=subprocess.PIPE, stdout=f)
+        f.write('\n')
 
-FILENAME=diags-`date +%Y%m%d_%H%M%S`.tar.gz
+        f.write("ip -6 route")
+        subprocess.Popen(["ip", "-6", "route"], stdin=subprocess.PIPE, stdout=f)
+        f.write('\n')
 
-tar -zcf $FILENAME *
-echo "Diags saved to $FILENAME in $diags_dir"
+    print("Dumping iptables")
+    with open(os.path.join(temp_diags_dir, 'iptables'), 'w') as f:
+        output = subprocess.Popen(["iptables-save"], stdin=subprocess.PIPE, stdout=f)
 
-echo "Uploading file. Available for 14 days from the URL printed when the upload completes"
-curl --upload-file $FILENAME https://transfer.sh/$FILENAME
-
-popd >/dev/null
+    print("Dumping ipset")
+    ipset_file = os.path.join(temp_diags_dir, 'ipset')
+    with open(ipset_file, 'w') as f:
+        try:
+            subprocess.Popen(["ipset", "list"], stdin=subprocess.PIPE, stdout=f)
+        except:
+            # System might not have ipset command. For now, we just ignore
+            print("WARNING: Couldn't dump ipset")
+
+    print("Copying Calico logs")
+    calico_dir = '/var/log/calico'
+    subprocess.Popen(["cp", "-a", calico_dir, temp_diags_dir])
+
+    print("Dumping datastore")
+    url = "http://127.0.0.1:4001/v2/keys/calico?recursive=true"
+    datastore_file = os.path.join(temp_diags_dir, 'etcd_calico')
+    curl_process = subprocess.Popen(["curl", "-s", "-L", url, "-o", datastore_file], stdout=subprocess.PIPE)
```

Comment:
Oh interesting... is there not an `etcdctl` command for this, or do we not trust that to work?